--- charmrun.c.orig	2009-04-23 14:46:33.000000000 -0700
+++ charmrun.c	2009-04-29 22:09:00.000000000 -0700
@@ -1322,7 +1322,7 @@ char *input_scanf_chars(fmt)
     char *tmp=tmpnam(NULL);/*This was once /tmp/fnord*/
 #endif
     unlink(tmp);
-    fd = open(tmp,O_RDWR | O_CREAT | O_TRUNC);
+    fd = open(tmp,O_RDWR | O_CREAT | O_TRUNC, 0666 );
     if (fd<0) { 
       fprintf(stderr,"cannot open temp file /tmp/fnord");
       exit(1);
