--- source/misc/pbeq.src.orig   2008-06-19 11:06:10.000000000 -0700
+++ source/misc/pbeq.src        2008-06-19 11:07:34.000000000 -0700
@@ -198,6 +198,11 @@
          QMMIJ     = .FALSE.
       ENDIF

+      ELSEIF(WRD.EQ.'APBS')THEN
+       WRITE(OUTU,100)
+       WRITE(OUTU,100) 'PBEQ> Entering APBS module'
+       call apbs
+
 C.......................................................................
       ELSEIF(WRD.EQ.'SOLV')THEN
 C            -------------
