--- install.com.orig	2008-06-19 10:28:20.000000000 -0700
+++ install.com	2008-06-19 10:28:43.000000000 -0700
@@ -651,8 +651,8 @@
 # setup for APBS
   if ( $apbs == 1 && $chm_host == gnu ) then
       setenv APBS Y
-      if ( ! $?APBS_LIB || ! $?IAPBS_LIB || ! $?APBS_BLAS ) then
-        echo "APBS_LIB, IAPBS_LIB and APBS_BLAS variables must be set!"
+      if ( ! $?APBS_LIB || ! $?APBS_BLAS ) then
+        echo "APBS_LIB and APBS_BLAS variables must be set!"
 	exit
       endif           
   endif
